### PR TITLE
JDBC refactoring + connection keep-alive

### PR DESCRIPTION
**Pull request by wendigo**
Minor refactoring of connection factories structure that makes it easier to introduce new connection factories without adding new annotations

### CODE REVIEW

**Comment by wendigo** [comment]
ping @ebyhr @hashhar

**Comment by mosabua** [comment]
I assume this is still in progress @wendigo ...

**Comment by wendigo** [comment]
Just a little of background here.

In Trino codebase, we prefer composition of behaviours, rather than inheritance, which gives more flexibility when we are extending some "provided"  behaviour, i.e. when JDBC connector extends base-jdbc one.

Usually, it leads to a collection of classes, that will wrap around each other, augmenting or changing behaviour of the delegate instance.

i.e. we have `JdbcClient` (interface), `BaseJdbcClient` (implementation), `CachingJdbcClient` (wrapper), `StatisticsAwareJdbcClient` (wrapper).

They are constructed and linked together through the series of bindings.

`@StatsCollecting StatisticsAwareJdbcClient(@ForBaseJdbc JdbcClient client)`
`CachingJdbcClient(@StatsCollecting delegate)`

This couples these two implementations together, so if I would like to introduce 
new behaviour that taps in between the `StatisticsAwareJdbcClient` and `CachingJdbcClient`
I need to introduce new annotation, change the `CachingJdbcClient` so it accepts this new annotation and ensure
that every down stream connector will correctly wire this new hierarchy of behaviours.

Instead of having this strong-coupling between implementations, I'd like to have it defined in terms of "which behaviour is the outer most and which ones are inner most"
by defining a soft order that will sort behaviours `[A (priority=1), C (priority=3), D (priority=5), B(priority=2), F (priority=100)]` to provide a final binding `A(B(C(D(F))))`.

This PR makes it possible to relax these strong-relationships between classes.

**Comment by wendigo** [comment]
For `ConnectionFactory` we have these behaviour atm:

- Statistics collection,
- Lazy connection initialization
- Retrying connection acquisition
- Reusable connection for metadata resolution


and bunch of custom `ConnectionFactories` that creates connection in a given connector for a specific authentication mode (they compose in some cases as well).

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ForwardingConnectionFactory.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+package io.trino.plugin.jdbc;
+
+import io.trino.spi.connector.ConnectorSession;
+import jakarta.annotation.PreDestroy;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+
+public abstract class ForwardingConnectionFactory
+        implements ConnectionFactory
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ForwardingConnectionFactory.java:23]: I don't see the benefit of ForwardingConnectionFactory. There is only one method to implement (besides close) and all subclasses will override it. Requiring explicit close is a plus IMO.

**Pull request closed without merging**